src/searching/binary_search.rs

```diff
@@ -1,106 +1,140 @@
+//! This module provides an implementation of a binary search algorithm that
+//! works for both ascending and descending ordered arrays. The binary search
+//! function returns the index of the target element if it is found, or `None`
+//! if the target is not present in the array.
+
 use std::cmp::Ordering;
 
+/// Performs a binary search for a specified item within a sorted array.
+///
+/// This function can handle both ascending and descending ordered arrays. It
+/// takes a reference to the item to search for and a slice of the array. If
+/// the item is found, it returns the index of the item within the array. If
+/// the item is not found, it returns `None`.
+///
+/// # Parameters
+///
+/// - `item`: A reference to the item to search for.
+/// - `arr`: A slice of the sorted array in which to search.
+///
+/// # Returns
+///
+/// An `Option<usize>` which is:
+/// - `Some(index)` if the item is found at the given index.
+/// - `None` if the item is not found in the array.
 pub fn binary_search<T: Ord>(item: &T, arr: &[T]) -> Option<usize> {
```

Comment:
General remark: each time I see a _binary search_ like things , which require sorted inputs, I am asking myself, why not to implement a `SortedArray` type and use them as inputs for such functions.

Reply:
I have reviewed your suggestions and would like to share my opinion. Integrating the `SortedArray` struct into the binary search algorithm will streamline the testing process by enabling the use of randomly-ordered arrays in test cases. The `SortedArray` struct will manage the sorting of arrays based on the `is_asc` argument passed into the tests.

Here are my suggestions that add the `SortedArray` struct into the `binary_search` algorithm along with test cases:
```rust
use std::cmp::Ordering;

#[derive(Debug, PartialEq, Eq)]
pub struct SortedArray<T: Ord> {
    data: Vec<T>,
    is_asc: bool,
}

impl<T: Ord> SortedArray<T> {
    /// Creates a new SortedArray from a Vec. The Vec is sorted in ascending or
    /// descending order based on the `is_asc` parameter.
    ///
    /// # Parameters
    ///
    /// - `mut data`: A mutable Vec of elements to be sorted.
    /// - `is_asc`: A boolean indicating whether the array should be
    /// sorted in ascending order (`True`) or descending order (`False`).
    ///
    /// # Returns
    ///
    /// A new SortedArray instance.
    pub fn new(mut data: Vec<T>, is_asc: bool) -> Self {
        match is_asc {
            true => data.sort(),
            false => data.sort_by(|a, b| b.cmp(a)),
        }

        Self { data, is_asc }
    }

    /// Returns a reference to the inner data slice.
    pub fn as_slice(&self) -> &[T] {
        &self.data
    }

    /// Returns the length of the sorted array.
    pub fn len(&self) -> usize {
        self.data.len()
    }

    /// Returns whether the array is sorted in ascending order.
    pub fn is_asc(&self) -> bool {
        self.is_asc
    }
}

/// Performs a binary search for a specified item within a SortedArray.
///
/// This function can handle both ascending and descending ordered arrays. It
/// takes a reference to the item to search for and a SortedArray. If the item
/// is found, it returns the index of the item within the array. If the item is
/// not found, it returns `None`.
///
/// # Parameters
///
/// - `item`: A reference to the item to search for.
/// - `arr`: A reference to the SortedArray in which to search.
///
/// # Returns
///
/// An `Option<usize>` which is:
/// - `Some(index)` if the item is found at the given index.
/// - `None` if the item is not found in the array.
pub fn binary_search<T: Ord>(item: &T, arr: &SortedArray<T>) -> Option<usize> {
    let is_asc = arr.is_asc();

    let mut left = 0;
    let mut right = arr.len();

    while left < right {
        if match_compare(item, arr, &mut left, &mut right, is_asc) {
            return Some(left);
        }
    }

    None
}

/// Compares the item with the middle element of the current search range and
/// updates the search bounds accordingly. This function handles both ascending
/// and descending ordered arrays. It calculates the middle index of the
/// current search range and compares the item with the element at
/// this index. It then updates the search bounds (`left` and `right`) based on
/// the result of this comparison. If the item is found, it updates `left` to
/// the index of the found item and returns `true`.
///
/// # Parameters
///
/// - `item`: A reference to the item to search for.
/// - `arr`: A reference to the SortedArray in which to search.
/// - `left`: A mutable reference to the left bound of the search range.
/// - `right`: A mutable reference to the right bound of the search range.
/// - `is_asc`: A boolean indicating whether the array is sorted in ascending order.
///
/// # Returns
///
/// A `bool` indicating whether the item was found.
fn match_compare<T: Ord>(
    item: &T,
    arr: &SortedArray<T>,
    left: &mut usize,
    right: &mut usize,
    is_asc: bool,
) -> bool {
    let mid = *left + (*right - *left) / 2;
    let cmp_result = item.cmp(&arr.as_slice()[mid]);

    match (is_asc, cmp_result) {
        (true, Ordering::Less) | (false, Ordering::Greater) => {
            *right = mid;
        }
        (true, Ordering::Greater) | (false, Ordering::Less) => {
            *left = mid + 1;
        }
        (_, Ordering::Equal) => {
            *left = mid;
            return true;
        }
    }

    false
}

#[cfg(test)]
mod tests {
    use super::*;

    const STR_ARR: &[&str] = &["b", "d", "c", "a", "zoo", "google"];
    const INT_ARR: &[i32] = &[3, 2, 4, 1];
    const INT_ARR_WITH_GAP: &[i32] = &[11, 3, 8, 1];

    macro_rules! test_cases {
        ($($name:ident: $test_case:expr,)*) => {
            $(
                #[test]
                fn $name() {
                    let (item, arr, is_asc, expected) = $test_case;
                    let sorted_array = SortedArray::new(arr.to_vec(), is_asc);
                    assert_eq!(binary_search(&item, &sorted_array), expected);
                }
            )*
        };
    }

    test_cases! {
        empty: ("a", &[] as &[&str], true, None),
        one_item_found: ("a", &["a"], true, Some(0)),
        one_item_not_found: ("b", &["a"], true, None),
        search_strings_asc_start: ("a", STR_ARR, true, Some(0)),
        search_strings_asc_middle: ("google", STR_ARR, true, Some(4)),
        search_strings_asc_last: ("zoo", STR_ARR, true, Some(5)),
        search_strings_asc_not_found: ("x", STR_ARR, true, None),
        search_strings_desc_start: ("zoo", STR_ARR, false, Some(0)),
        search_strings_desc_middle: ("google", STR_ARR, false, Some(1)),
        search_strings_desc_last: ("a", STR_ARR, false, Some(5)),
        search_strings_desc_not_found: ("x", STR_ARR, false, None),
        search_ints_asc_start: (1, INT_ARR, true, Some(0)),
        search_ints_asc_middle: (3, INT_ARR, true, Some(2)),
        search_ints_asc_end: (4, INT_ARR, true, Some(3)),
        search_ints_asc_not_found: (5, INT_ARR, true, None),
        search_ints_desc_start: (4, INT_ARR, false, Some(0)),
        search_ints_desc_middle: (3, INT_ARR, false, Some(1)),
        search_ints_desc_end: (1, INT_ARR, false, Some(3)),
        search_ints_desc_not_found: (5, INT_ARR, false, None),
        with_gaps_0: (0, INT_ARR_WITH_GAP, true, None),
        with_gaps_1: (1, INT_ARR_WITH_GAP, true, Some(0)),
        with_gaps_2: (2, INT_ARR_WITH_GAP, true, None),
        with_gaps_3: (3, INT_ARR_WITH_GAP, true, Some(1)),
        with_gaps_4: (4, INT_ARR_WITH_GAP, true, None),
        with_gaps_5: (5, INT_ARR_WITH_GAP, true, None),
        with_gaps_6: (6, INT_ARR_WITH_GAP, true, None),
        with_gaps_7: (7, INT_ARR_WITH_GAP, true, None),
        with_gaps_8: (8, INT_ARR_WITH_GAP, true, Some(2)),
        with_gaps_9: (9, INT_ARR_WITH_GAP, true, None),
        with_gaps_10: (10, INT_ARR_WITH_GAP, true, None),
        with_gaps_11: (11, INT_ARR_WITH_GAP, true, Some(3)),
        with_gaps_12: (12, INT_ARR_WITH_GAP, true, None),
        with_gaps_13: (13, INT_ARR_WITH_GAP, true, None),
    }
}
```

Feel free to give suggestions to improve this implementation. So that we can have a unique implementation that is the best suited for `binary_search`

Reply:
This current implementation passed all the tests provided.

Reply:
`SortedArray` itself could contain information how is it sorted. Besides that, feel free to create another PR with such suggestion. Clearly it would be great to apply it in other places, where sorted input is needed.

Another thing, which one needs to think are duplicates.

Reply:
We should implement the suggested changes in this pull request, shouldn't we?

Reply:
> `SortedArray` itself could contain information how is it sorted. Besides that, feel free to create another PR with such suggestion. Clearly it would be great to apply it in other places, where sorted input is needed. What I mean is that there should be available types like `SortedArray::<i32, Ascending>` and `SortedArray::<i32, Descending>`.
> 
> Another thing, which one needs to think are duplicates.

"Thank you for your suggestions. I will take a look at them."

Reply:
Maybe one could also introduce things like `SortedArray::<i32, StrictlyAscending>` and `SortedArray::<i32, StrictlyDescending>`.

In any case: it will be a fun project. Please create a PR with such class applied only in `binary_search`. Then we will see.

Reply:
Before creating a pull request (PR), I will share my proposal to address the remaining issues with binary search:
- Create a trait to define the comparison behavior based on the ordering type.
- When handling duplicates, we should return all occurrences of the searched item. This may add some complexity depending on the number of duplicated items.

Here is my suggestion:

```rs
//! This module provides an implementation of a binary search algorithm that
//! works for both ascending and descending ordered arrays. The binary search
//! function returns the index of the target element if it is found, or `None`
//! if the target is not present in the array.

use std::cmp::Ordering;

/// Marker type for ascending order.
pub struct Ascending;

/// Marker type for descending order.
pub struct Descending;

#[derive(Debug, PartialEq, Eq)]
pub struct SortedArray<T: Ord, O> {
    data: Vec<T>,
    _order: std::marker::PhantomData<O>,
}

impl<T: Ord, O> SortedArray<T, O> {
    /// Returns a reference to the inner data slice.
    pub fn as_slice(&self) -> &[T] {
        &self.data
    }

    /// Returns the length of the sorted array.
    pub fn len(&self) -> usize {
        self.data.len()
    }
}

impl<T: Ord> SortedArray<T, Ascending> {
    /// Creates a new SortedArray from a Vec in ascending order.
    ///
    /// # Parameters
    ///
    /// - `mut data`: A mutable Vec of elements to be sorted.
    ///
    /// # Returns
    ///
    /// A new SortedArray instance.
    pub fn new(mut data: Vec<T>) -> Self {
        data.sort();
        Self {
            data,
            _order: std::marker::PhantomData,
        }
    }
}

impl<T: Ord> SortedArray<T, Descending> {
    /// Creates a new SortedArray from a Vec in descending order.
    ///
    /// # Parameters
    ///
    /// - `mut data`: A mutable Vec of elements to be sorted.
    ///
    /// # Returns
    ///
    /// A new SortedArray instance.
    pub fn new(mut data: Vec<T>) -> Self {
        data.sort_by(|a, b| b.cmp(a));
        Self {
            data,
            _order: std::marker::PhantomData,
        }
    }
}

/// Performs a binary search for a specified item within a SortedArray,
/// returning a vector of all occurrences of the item if duplicates exist.
///
/// This function can handle both ascending and descending ordered arrays.
/// It takes a reference to the item to search for and a SortedArray.
/// If the item is found, it returns a vector of indices of the item within the array.
/// If the item is not found, it returns an empty vector.
///
/// # Parameters
///
/// - `item`: A reference to the item to search for.
/// - `arr`: A reference to the SortedArray in which to search.
///
/// # Returns
///
/// A `Vec<usize>` containing all indices of the item if found, otherwise an empty vector.
pub fn binary_search<T: Ord, O>(item: &T, arr: &SortedArray<T, O>) -> Vec<usize>
where
    O: OrdOrder,
{
    let mut left = 0;
    let mut right = arr.len();
    let mut result = Vec::new();

    while left < right {
        let mid = left + (right - left) / 2;
        match O::cmp(item, &arr.as_slice()[mid]) {
            Ordering::Less => right = mid,
            Ordering::Greater => left = mid + 1,
            Ordering::Equal => {
                // Collect all occurrences
                let mut index = mid;
                while index > 0 && arr.as_slice()[index - 1] == *item {
                    index -= 1;
                }
                while index < arr.len() && arr.as_slice()[index] == *item {
                    result.push(index);
                    index += 1;
                }
                break;
            }
        }
    }

    result
}

/// Trait to define the comparison behavior based on the ordering type.
pub trait OrdOrder {
    fn cmp<T: Ord>(a: &T, b: &T) -> Ordering;
}

impl OrdOrder for Ascending {
    fn cmp<T: Ord>(a: &T, b: &T) -> Ordering {
        a.cmp(b)
    }
}

impl OrdOrder for Descending {
    fn cmp<T: Ord>(a: &T, b: &T) -> Ordering {
        b.cmp(a)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const STR_ARR: &[&str] = &["b", "d", "c", "a", "zoo", "google"];
    const INT_ARR: &[i32] = &[3, 2, 4, 1];
    const INT_ARR_WITH_GAP: &[i32] = &[11, 3, 8, 1];
    const DUPLICATE_ARR: &[i32] = &[1, 2, 5, 3, 2, 3, 3, 4, 4, 5];

    macro_rules! test_empty {
        ($($name:ident: $test_case:expr,)*) => {
            $(
                #[test]
                fn $name() {
                    let (item, arr) = $test_case;
                    let sorted_array = SortedArray::<_, Ascending>::new(arr.to_vec());
                    assert_eq!(binary_search(&item, &sorted_array), Vec::new());
                }
            )*
        };
    }

    macro_rules! test_one_item {
        ($($name:ident: $test_case:expr,)*) => {
            $(
                #[test]
                fn $name() {
                    let (item, data, expected) = $test_case;
                    let sorted_array = SortedArray::<_, Ascending>::new(vec![data]);
                    assert_eq!(binary_search(&item, &sorted_array), expected);
                }
            )*
        };
    }

    macro_rules! test_search_cases {
        ($($name:ident: $test_case:expr,)*) => {
            $(
                #[test]
                fn $name() {
                    let (item, data, expected_asc, expected_desc) = $test_case;

                    let sorted_array = SortedArray::<_, Ascending>::new(data.to_vec());
                    assert_eq!(binary_search(&item, &sorted_array), expected_asc);

                    let sorted_array = SortedArray::<_, Descending>::new(data.to_vec());
                    assert_eq!(binary_search(&item, &sorted_array), expected_desc);
                }
            )*
        };
    }

    macro_rules! test_duplicates {
        ($($name:ident: $test_case:expr,)*) => {
            $(
                #[test]
                fn $name() {
                    let (item, data, expected) = $test_case;
                    let sorted_array = SortedArray::<_, Ascending>::new(data.to_vec());
                    assert_eq!(binary_search(&item, &sorted_array), expected);
                }
            )*
        };
    }

    test_empty! {
        empty_str: ("a", &[] as &[&str]),
        empty_int: (1, &[] as &[i32]),
    }

    test_one_item! {
        one_item_found_str: ("a", "a", vec![0]),
        one_item_not_found_str: ("b", "a", Vec::new()),
        one_item_found_int: (1, 1, vec![0]),
        one_item_not_found_int: (2, 1, Vec::new()),
    }

    test_search_cases! {
        search_strings_start: ("a", STR_ARR, vec![0], vec![5]),
        search_strings_middle: ("google", STR_ARR, vec![4], vec![1]),
        search_strings_last: ("zoo", STR_ARR, vec![5], vec![0]),
        search_strings_not_found: ("x", STR_ARR, Vec::new(), Vec::new()),
        search_ints_start: (1, INT_ARR, vec![0], vec![3]),
        search_ints_middle: (3, INT_ARR, vec![2], vec![1]),
        search_ints_end: (4, INT_ARR, vec![3], vec![0]),
        search_ints_not_found: (5, INT_ARR, Vec::new(), Vec::new()),
        with_gaps_0: (0, INT_ARR_WITH_GAP, Vec::new(), Vec::new()),
        with_gaps_1: (1, INT_ARR_WITH_GAP, vec![0], vec![3]),
        with_gaps_2: (2, INT_ARR_WITH_GAP, Vec::new(), Vec::new()),
        with_gaps_3: (3, INT_ARR_WITH_GAP, vec![1], vec![2]),
        with_gaps_4: (4, INT_ARR_WITH_GAP, Vec::new(), Vec::new()),
        with_gaps_5: (5, INT_ARR_WITH_GAP, Vec::new(), Vec::new()),
        with_gaps_6: (6, INT_ARR_WITH_GAP, Vec::new(), Vec::new()),
        with_gaps_7: (7, INT_ARR_WITH_GAP, Vec::new(), Vec::new()),
        with_gaps_8: (8, INT_ARR_WITH_GAP, vec![2], vec![1]),
        with_gaps_9: (9, INT_ARR_WITH_GAP, Vec::new(), Vec::new()),
        with_gaps_10: (10, INT_ARR_WITH_GAP, Vec::new(), Vec::new()),
        with_gaps_11: (11, INT_ARR_WITH_GAP, vec![3], vec![0]),
        with_gaps_12: (12, INT_ARR_WITH_GAP, Vec::new(), Vec::new()),
        with_gaps_13: (13, INT_ARR_WITH_GAP, Vec::new(), Vec::new()),
    }

    test_duplicates! {
        search_duplicates_first: (2, DUPLICATE_ARR, vec![1, 2]),
        search_duplicates_middle: (3, DUPLICATE_ARR, vec![3, 4, 5]),
        search_duplicates_end: (5, DUPLICATE_ARR, vec![8, 9]),
        search_duplicates_not_found: (6, DUPLICATE_ARR, Vec::new()),
    }
}
```

The result of binary search will now always be vector. If you see this proposal has the potential to be a PR, please let me know.